### PR TITLE
[RPD-273] Improvements to the `provision` user experience

### PR DESCRIPTION
**Pull request by JonoCX**
While `provision` is running, there isn't much output to the user besides the spinner. This is great, it gives them a sense of time but there's no context. In this PR, we add some output to the provision process to improve the experience for the user.

There's a set of facts in the backend about the stack and Matcha (the drink) which are randomly selected and displayed to the user at a regular interval. For this to work, a separate thread is created while the `apply` call is running. This is then terminated once Terraform has finished.

This PR also introduces `TerraformResult` which is an object containing the outcome of the calls to the Terraform Python library.

## Checklist

Please ensure you have done the following:

* [x] I have read the [CONTRIBUTING](https://github.com/fuzzylabs/matcha/blob/main/CONTRIBUTING.md) guide.
* [x] I have updated the documentation if required.
* [x] I have added tests which cover my changes.

## Type of change

Tick all those that apply:

* [ ] Bug Fix (non-breaking change, fixing an issue)
* [x] New feature (non-breaking change to add functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Other (add details above)

### CODE REVIEW

**Review by Christopher-Norman — COMMENTED** [review]
Looking good so far, are the `INFRA_FACTS` meant to stack?:
```
✔ Matcha 🍵 initialized!

Samurai's drank Matcha before battles
Did you know that Matcha tea was created by accident?
Everything being provisioned is fully open source
Everything being provisioned is fully open source
The brewing temperature of the affects the taste of Matcha
Samurai's drank Matcha before battles
⠦ 0:02:59 Applying
```

File: src/matcha_ml/services/analytics_service.py [hunk 1]
```diff
@@ -27,12 +27,14 @@ class AnalyticsEvent(str, Enum):
 
 
 def execute_analytics_event(
-    func: Callable, *args, **kwargs
+    func: Callable, *args: dict, **kwargs: dict  # type: ignore
```
Comment (KirsoppJ) [inline src/matcha_ml/services/analytics_service.py:30]: not sure what's going on with the type hints here, but should *args really be a dict not a list?

**Review by KirsoppJ — COMMENTED** [review]
Looking good so far, haven't run it end to end yet but will before approving

**Review by Christopher-Norman — APPROVED** [review]
LGTM

File: src/matcha_ml/cli/ui/spinner.py [hunk 2]
```diff
@@ -30,9 +30,14 @@ def __init__(self, status: str):
         )
         self.progress.add_task(description=status, total=None)
 
-    def __enter__(self) -> None:
-        """Call when a spinner object is created using a `with` statement."""
+    def __enter__(self):  # type: ignore
```
Comment (KirsoppJ) [inline src/matcha_ml/cli/ui/spinner.py:33]: Nit: Just a thought - you could type this return with TypeVar or with a string.

**Review by KirsoppJ — APPROVED** [review]

File: src/matcha_ml/cli/ui/spinner.py (hunk 2 above)
Comment (JonoCX) [reply src/matcha_ml/cli/ui/spinner.py:33]: Good idea, I'll give it a try 👍🏻

**Review by KirsoppJ — APPROVED** [review]
Looks great!

**Comment by strickvl** [comment]
Love that you ran with this idea! 🙌

**Pull request merged**